packages/web/src/components/composer-input/ComposerInput.tsx

```diff
@@ -0,0 +1,363 @@
+  } = useAudiusLinkResolver({
+    value,
+    hostname: env.PUBLIC_HOSTNAME,
+    audiusSdk
+  })
+
+  useEffect(() => {
+    const fn = async () => {
+      if (presetMessage) {
+        const editedValue = await resolveLinks(presetMessage)
+        setValue(editedValue)
+      }
+    }
+    fn()
+  }, [presetMessage, resolveLinks])
+
+  useEffect(() => {
+    onChange?.(restoreLinks(value), linkEntities)
+  }, [linkEntities, onChange, restoreLinks, value])
+
+  useEffect(() => {
+    if (messageId !== messageIdRef.current) {
+      messageIdRef.current = messageId
+      setValue('')
+    }
+  }, [messageId])
+
+  // useEffect(() => {
+  //   if (!focused && isUserAutocompleteActive) {
+  //     setIsUserAutocompleteActive(false)
+  //   }
+  // }, [focused, isUserAutocompleteActive])
+
+  const getUserMentions = useCallback(
+    (value: string) => {
+      const regexString = [...userMentions]
+        .sort((a, b) => b.length - a.length)
+        .join('|')
+      const regex = regexString.length ? new RegExp(regexString, 'g') : null
+
+      return regex
+        ? Array.from(value.matchAll(regex)).map((match) => ({
+            type: 'mention',
+            text: match[0],
+            index: match.index,
+            link: ''
+          }))
+        : null
+    },
+    [userMentions]
+  )
+
+  const handleChange = useCallback(
+    async (e: ChangeEvent<HTMLTextAreaElement>) => {
+      const textarea = e.target as HTMLTextAreaElement
+      const cursorPosition = textarea.selectionStart
+      setValue(e.target.value)
+      const editedValue = await resolveLinks(e.target.value, cursorPosition)
+      setValue(editedValue)
+      // TODO: Need to update this to move to the proper position affect link change to human text
+      // setTimeout(() => {
+      //   textarea.selectionStart = cursorPosition
+      //   textarea.selectionEnd = cursorPosition
+      // }, 0)
+    },
+    [resolveLinks, setValue]
+  )
+
+  const handleSubmit = useCallback(() => {
+    onSubmit?.(restoreLinks(value), linkEntities)
+  }, [linkEntities, onSubmit, restoreLinks, value])
+
+  // Submit when pressing enter while not holding shift
+  const handleKeyDown = useCallback(
+    (e: React.KeyboardEvent<HTMLTextAreaElement>) => {
+      if (isUserAutocompleteActive) {
+        if (e.key === ENTER_KEY || e.key === TAB_KEY) {
+          e.preventDefault()
+          // TODO: Fill in the top result
+        }
+
+        if (e.key === ESCAPE_KEY || e.key === SPACE_KEY) {
+          setIsUserAutocompleteActive(false)
+        }
+
+        if (e.key === BACKSPACE_KEY) {
+          const textarea = e.target as HTMLTextAreaElement
+          const cursorPosition = textarea.selectionStart
+          const deletedChar = textarea.value[cursorPosition - 1]
+          if (deletedChar === '@') {
+            setIsUserAutocompleteActive(false)
+          }
+        }
+
+        return
+      }
+
+      // Submit on enter
+      if (e.key === ENTER_KEY && !e.shiftKey) {
```

Comment:
why the specific exclusion of Shift? I'm not really following

Reply:
on shift+enter a new line is entered

Reply:
Oh I see, for some reason I thought this was key handling inside the autocomplete window. Mb